src/state-item.c

```diff
@@ -0,0 +1,417 @@
+            if (item_number_is_rule_number(item) || ISTOKEN(item)) continue;
+            symbol_number sym = item_number_as_symbol_number(item);
+            bitset lb = hash_pair_lookup(closure_map, sym);
+            if (lb != NULL) {
+                hash_pair *prod_hp = xmalloc(sizeof(hash_pair));
+                prod_hp->key = j;
+                prod_hp->l = lb;
+                //update prods
+                if (!hash_insert(prods, prod_hp))
+                    xalloc_die();
+                
+                //update rev_prods
+                bitset_iterator biter;
+                state_item_number prod;
+                BITSET_FOR_EACH (biter, lb, prod, 0) {
+                    bitset rev_itms = rev_prods_lookup(prod);
+                    if (rev_itms == NULL) {
+                        rev_itms = bitset_create(nstate_items, BITSET_SPARSE);
+                        hash_pair_insert(rev_prods, prod, rev_itms);
+                    }
+                    bitset_set(rev_itms, j);
+                }
+            }
+        }
+        
+    }
+}
+
+/* Since lookaheads are only generated for reductions,
+   we need to propogate lookahead sets backwards as
+   the searches require each state_item to have a lookahead.
+ */
+static inline void
+gen_lookaheads()
+{
+    for (state_item_number i = 0; i < nstate_items; i++)
+    {
+        state_item *si = state_items + i;
+        if (item_number_is_symbol_number(*(si->item)) || si->state->reductions->lookahead_tokens == NULL)
+            continue;
+        
+        bitset lookahead = si->state->reductions->lookahead_tokens[si->reduction_offset];
+        gl_list_t queue = gl_list_create(GL_LINKED_LIST, NULL, NULL, NULL, true, 1, (const void **)&si);
+        
+        // For each reduction item, traverse through all state_items
+        // accessible through reverse transition steps, and set their
+        // lookaheads to the reduction items lookahead
+        while (gl_list_size(queue) > 0)
+        {
+            state_item *prev = gl_list_get_at(queue, 0);
+            gl_list_remove_at(queue, 0);
+            prev->lookahead = lookahead;
+            bitset rsi = rev_trans[prev->num];
+            bitset_iterator biter;
+            state_item_number sin;
+            BITSET_FOR_EACH (biter, rsi, sin, 0)
+                gl_list_add_last(queue, &state_items[sin]);
+        }
+    }
+}
+
+void
+print_state_item(state_item *si, FILE *out)
+{
+    fprintf(out, "%d:", si->state->number);
+    item_print(si->item, NULL, out);
+    putc('\n', out);
+}
+
+/**
+ * Report set counts and the state_item graph if trace is enabled
+ */
+void
+state_items_report(FILE *out)
+{
+    state_items_init(out);
+    fprintf(out, "# state items: %zu\n", nstate_items);
+    
+    size_t count = 0;
+    for (state_item_number si = 0; si < nstate_items; si++)
+        if (trans[si] != -1)
+            count++;
+    
+    fprintf(out, "# transitions: %zu\n", count);
+    
+    count = 0;
```

Comment:
Instead of reusing the same variable, please be clearer and use scopes.  And avoid `size_t` and other unsigned types when you can (we are in the process of migrating away from unsigned types, there might be places in Bison that are still using them).

``` C
  {
     int count = 0;
     for (state_item_number si = 0; si < nstate_items; ++si)
         count += trans[si] != -1;
     fprintf (out, "# transitions: %d\n", count);
  }
```